core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java

```diff
@@ -239,77 +240,152 @@
+     *                                      any annotation. If it is {@code true}, a list of
+     *                                      {@link RequestConverterFunction} would be applied for creating an
+     *                                      object for the element. Otherwise, the element would be ignored.
      */
     private static Optional<AnnotatedValueResolver> of(AnnotatedElement annotatedElement,
                                                        AnnotatedElement typeElement, Class<?> type,
                                                        Set<String> pathParams,
-                                                       List<RequestObjectResolver> objectResolvers) {
+                                                       List<RequestConverterFunction> converters,
+                                                       boolean applyConverterForNoAnnotation) {
         requireNonNull(annotatedElement, "annotatedElement");
         requireNonNull(typeElement, "typeElement");
         requireNonNull(type, "type");
         requireNonNull(pathParams, "pathParams");
-        requireNonNull(objectResolvers, "objectResolvers");
+        requireNonNull(converters, "objectResolvers");
 
         final Param param = annotatedElement.getAnnotation(Param.class);
         if (param != null) {
             final String name = findName(param, typeElement);
             if (pathParams.contains(name)) {
-                return Optional.of(ofPathVariable(param, name, annotatedElement, typeElement, type));
+                return Optional.of(ofPathVariable(name, annotatedElement, typeElement, type));
             } else {
-                return Optional.of(ofHttpParameter(param, name, annotatedElement, typeElement, type));
+                return Optional.of(ofHttpParameter(name, annotatedElement, typeElement, type));
             }
         }
 
         final Header header = annotatedElement.getAnnotation(Header.class);
         if (header != null) {
-            return Optional.of(ofHeader(header, annotatedElement, typeElement, type));
+            return Optional.of(ofHeader(findName(header, typeElement), annotatedElement, typeElement, type));
         }
 
+        if (type == Optional.class) {
+            // Do not allow Optional<?> for the following auto-injectable types:
+            //  - RequestContext and ServiceRequestContext
+            //  - Request and HttpRequest
+            //  - AggregatedHttpMessage
+            //  - HttpParameters
+            //  - Cookies
+            final Type parameterizedType = parameterizedTypeOf(annotatedElement);
+            assert parameterizedType instanceof ParameterizedType : String.valueOf(parameterizedType);
+            final Type actual = ((ParameterizedType) parameterizedType).getActualTypeArguments()[0];
+            if (autoInjectableTypes.containsKey(actual)) {
+                return Optional.empty();
```

Comment:
Don't we have to throw an exception because we do not allow Optional<?> for auto-injectable types?

Reply:
Changed to throw IAE. Thanks!